XWOpt/OptReader.cs

```diff
@@ -0,0 +1,310 @@
+                        default:
+                            logger("Found unknown node type " + majorId + " " + minorId + " at " + BaseStream.Position + " context:" + context);
+                            return new Texture(this, preHeaderOffset);
+                    }
+                default:
+                    return new BaseNode(this);
+            }
+        }
+
+        /// <summary>
+        /// Calculate actual file address based on a pointer stored in the file.
+        /// </summary>
+        /// <param name="offset">Position read from file</param>
+        /// <returns>Physical file address</returns>
+        internal long RealOffset(int offset)
+        {
+            return offset - globalOffset;
+        }
+
+        /// <summary>
+        /// Calculates the pointer address that would be stored in a file based on actual file address.
+        /// </summary>
+        /// <param name="offset">Real position in file</param>
+        /// <returns>Position to be stored.</returns>
+        internal long FakeOffset(int offset)
+        {
+            return offset + globalOffset;
+        }
+
+        /// <summary>Follow a pointer read from the file.</summary>
+        internal void FollowPointer()
+        {
+            Seek(ReadInt32());
+        }
+
+        /// <summary>Follow offset address in file. Warn if seek does not lead to next byte after address's location.</summary>
+        /// <remarks>This is useful for detecting unexpected data in the file.</remarks>
+        internal void FollowPointerToNextByte(object caller)
+        {
+            int offset = ReadInt32();
+            Seek(offset);
+            if (RealOffset(offset) != BaseStream.Position)
+            {
+                logger(String.Format("Warning: skipping unexpected {0} bytes at {1:X} in a {2}", RealOffset(offset) - BaseStream.Position, BaseStream.Position, caller.ToString()));
+            }
+        }
+
+        /// <summary>Set stream position accounting for global offset mechanism</summary>
+        internal void Seek(int offset)
+        {
+            BaseStream.Seek(RealOffset(offset), SeekOrigin.Begin);
+        }
+
+        internal void SeekShouldPointHere(int offset, object caller)
+        {
+            if (RealOffset(offset) != BaseStream.Position)
+            {
+                logger(String.Format(CultureInfo.CurrentCulture, "Warning: Skipping unkown data near {0:X} ({1} bytes) in a {2}", BaseStream.Position, RealOffset(offset) - BaseStream.Position, caller.ToString()));
+                Seek(offset);
+            }
+        }
+
+        /// <summary>
+        /// Reads an int for which we don't know what the value is supposed to be used for.
+        /// Logs a warning if it does not contain the value we normally see.
+        /// </summary>
+        /// <param name="expected">The expected value</param>
+        internal void ReadUnknownUseValue(int expected, object context)
```

Comment:
Are there ever cases where we have no idea what should come next?  Is it worth having a special case for these?  For example, if expected == null, then we don't bother logging a 'failure', simply because we don't care about the value and we have no idea what it should be anyway?

Reply:
The problem is that there are a lot of fields where I don't know the semantics of possible values.  I'm using `ReadUnknownUseValue()` and friends to sniff out different use cases.  Hopefully I'll be able to eventually reverse engineer these.  When I think I've figure out what a field is for, I'll change the caller to actually handle the semantics and throw an exception if it's out of bounds.